### PR TITLE
DHIS2-10712 - part 01 - updated duplicate checking on save

### PR DESCRIPTION
**Pull request by JoakimSM**
[DHIS2-10712](https://jira.dhis2.org/browse/DHIS2-10712)

First part of this ticket. Added new logic for duplicate checking on save. The next part will clean up the old code.

EDIT: some tests are failing, fixed in the next PR in this chain.

### CODE REVIEW

**Comment by paschalidi** [comment]
@JoakimSM you might wanna have a look  since linting is failing in this PR

**Comment by JoakimSM** [comment]
@paschalidi Fixed the linting problem, will resolve conflicts after we've merged your v36 PRs (up to you if you want to wait reviewing until conflicts are fixed)

**Comment by paschalidi** [comment]
FYI, I start with this review now

File: src/core_modules/capture-core/components/DataEntryUtils/duplicateChecker/checkForDuplicate.epic.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    return searchFoundation.convertValues(
+        formValuesForSearch,
+        (value, type) => convertClientToServer(convertFormToClient(value, type), type),
+    );
+};
+
+export const checkForDuplicateEpic = (action$: InputObservable, store: ReduxStore) =>
+    action$.pipe(
+        ofType(checkForDuplicateActionTypes.DUPLICATE_CHECK),
+        concatMap(({ payload: { searchGroup, dataEntryId, searchContext } }) => {
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/DataEntryUtils/duplicateChecker/checkForDuplicate.epic.js:30]: hm, why do we do concat map here?

File: src/core_modules/capture-core/components/Pages/NewRelationship/RegisterTei/RegisterTei.types.js [hunk 2]
```diff
@@ -17,4 +14,4 @@ export type OwnProps = {|
     onSave: Function,
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/Pages/NewRelationship/RegisterTei/RegisterTei.types.js:14]: A bit confused with this `onSave` function I see in some places we are changing the annotation of this function to `() => void`. However on the `RegisterTei.component.js` line 85 I can see that the function take 2 arguments. 

Are these two different onSave functions? If they are different perhaps we can find a way to make this more clear (?) 

If they arent different perhaps we could be consistent about its annotation?

File: src/core_modules/capture-core/components/Pages/NewRelationship/RegisterTei/DataEntry/TrackedEntityInstance/DataEntryTrackedEntityInstance.component.js [hunk 3]
```diff
@@ -1,23 +1,20 @@
...
 import teiClasses from './trackedEntityInstance.module.css';
 import { TeiRegistrationEntry } from '../../../../../DataEntries';
 import { useSaveButtonText } from '../useSaveButtonText';
-
-type Props = {|
-    theme: Theme,
-    onSave: () => void,
-    teiRegistrationMetadata?: TeiRegistration,
-|};
+import type { Props } from './dataEntryTrackedEntityInstance.types';
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/Pages/NewRelationship/RegisterTei/DataEntry/TrackedEntityInstance/DataEntryTrackedEntityInstance.component.js:8]: Cool!

**Review by paschalidi — COMMENTED** [review]
Hey, lots of good improvements here. I got few comments on this and one of a general comment regarding the file structure, see below.  Other than that looks solid

I am wondering if the `components/DataEntryUtils/**/*` items could be part of the `components/DataEntries` folder. For me it could either be part of the `components/DataEntries/common` or be its own `components/DataEntries/utils`. 

In my mind these utils will be only used for in the DataEntries, however let me know if I am missing some use case here.

**Review by JoakimSM — COMMENTED** [review]
Thanks for the review @paschalidi

> I am wondering if the components/DataEntryUtils/**/* items could be part of the components/DataEntries folder. For me it could either be part of the components/DataEntries/common or be its own components/DataEntries/utils.

> In my mind these utils will be only used for in the DataEntries, however let me know if I am missing some use case here.

I see why you could make a case for something like components/DataEntries/utils. But in this instance I think it makes sense to separate the DataEntry utility methods from the usage of the utility methods, especially if not all DataEntries will sit in the DataEntries folder (also in the future we would potentially expose the DataEntry utils for third parties and I think they are better exposed from something like DataEntryUtils).

Related, right now, the DataEntry folder itself exposes a bunch of utility methods (HOCs mostly) that I would like to move to DataEntryUtils (JIRA-ticket:  https://jira.dhis2.org/browse/TECH-583)

File: src/core_modules/capture-core/components/DataEntries/EnrollmentRegistrationEntry/EnrollmentRegistrationEntry.types.js [hunk 4]
```diff
@@ -1,27 +1,38 @@
...
   saveButtonText: string,
   fieldOptions?: Object,
-  onSave: (dataEntryId: string, itemId: string, formFoundation: RenderFoundation) => void,
-|}>
-
-export type HOCProps = {|
-  onPostProcessErrorMessage: Function,
-  ...CssClasses,
-|}
+  onSave: () => void,
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/DataEntries/EnrollmentRegistrationEntry/EnrollmentRegistrationEntry.types.js:12]: Is this onSave a different one?

File: src/core_modules/capture-core/components/DataEntries/TeiRegistrationEntry/TeiRegistrationEntry.types.js [hunk 5]
```diff
@@ -1,28 +1,38 @@
...
   saveButtonText: string,
   fieldOptions?: Object,
-  onSave: (dataEntryId: string, itemId: string, formFoundation: RenderFoundation) => void,
-|}>
-
-export type HOCProps = {|
-  onPostProcessErrorMessage: Function,
-  ...CssClasses,
-|}
+  onSave: () => void,
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/DataEntries/TeiRegistrationEntry/TeiRegistrationEntry.types.js:12]: +1

File: src/core_modules/capture-core/components/DataEntries/common/TEIAndEnrollment/DuplicateCheckOnSave/useDuplicateCheckerOnSave.types.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// @flow
+import { type InputSearchGroup, typeof scopeTypes } from '../../../../../metaData';
+
+export type Input = {|
+    onSave: () => void,
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/DataEntries/common/TEIAndEnrollment/DuplicateCheckOnSave/useDuplicateCheckerOnSave.types.js:5]: +1

File: src/core_modules/capture-core/components/DataEntries/TeiRegistrationEntry/TeiRegistrationEntry.types.js [hunk 7]
```diff
@@ -1,28 +1,38 @@
...
+    ...ContainerProps
+  |}>;
+
+type PropsAddedInHOC = {|
+    onPostProcessErrorMessage: Function,
+    ...CssClasses,
+|};
+type PropsRemovedInHOC = {|
+    renderDuplicatesCardActions?: RenderCustomCardActions,
+    renderDuplicatesDialogActions?: (onCancel: () => void, onSave: () => void) => Node,
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/DataEntries/TeiRegistrationEntry/TeiRegistrationEntry.types.js:33]: +1

File: src/core_modules/capture-core/components/DataEntries/common/TEIAndEnrollment/DuplicateCheckOnSave/withDuplicateCheckOnSave.types.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+// @flow
+import type { Node } from 'react';
+import type { Enrollment, TeiRegistration } from '../../../../../metaData';
+import type { RenderCustomCardActions } from '../../../../CardList';
+
+export type Props = {
+    id: string,
+    selectedScopeId: string,
+    onSave: () => void,
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/DataEntries/common/TEIAndEnrollment/DuplicateCheckOnSave/withDuplicateCheckOnSave.types.js:9]: +1

File: src/core_modules/capture-core/components/Pages/NewRelationship/RegisterTei/DataEntry/Enrollment/dataEntryEnrollment.types.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+// @flow
+import type { Node } from 'react';
+import type { Enrollment } from '../../../../../../metaData';
+import type { RenderCustomCardActions } from '../../../../../CardList';
+
+export type Props = {|
+    theme: Theme,
+    programId: string,
+    enrollmentMetadata?: Enrollment,
+    onSave: () => void,
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/Pages/NewRelationship/RegisterTei/DataEntry/Enrollment/dataEntryEnrollment.types.js:10]: +1

File: src/core_modules/capture-core/components/Pages/NewRelationship/RegisterTei/DataEntry/TrackedEntityInstance/dataEntryTrackedEntityInstance.types.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+// @flow
+import type { Node } from 'react';
+import type { TeiRegistration } from '../../../../../../metaData';
+import type { RenderCustomCardActions } from '../../../../../CardList';
+
+export type Props = {|
+    theme: Theme,
+    onSave: () => void,
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/Pages/NewRelationship/RegisterTei/DataEntry/TrackedEntityInstance/dataEntryTrackedEntityInstance.types.js:8]: +1

**Review by paschalidi — COMMENTED** [review]
Let me know if my question makes sense

**Comment by JoakimSM** [comment]
> Let me know if my question here #1603 (comment) makes sense
Thanks for the comments @paschalidi. I've improved upon the annotations now, it's not perfect, but better than it was. Please have a look.

File: src/core_modules/capture-core/components/DataEntries/index.js [hunk 11]
```diff
@@ -24,3 +24,4 @@ export {
 export { EnrollmentRegistrationEntry } from './EnrollmentRegistrationEntry/EnrollmentRegistrationEntry.container';
 export { TeiRegistrationEntry } from './TeiRegistrationEntry/TeiRegistrationEntry.container';
 export { SingleEventRegistrationEntry } from './SingleEventRegistrationEntry/SingleEventRegistrationEntry.container';
+export type { SaveForDuplicateCheck as SaveForEnrollmentAndTeiRegistration } from './common/TEIAndEnrollment/DuplicateCheckOnSave';
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/DataEntries/index.js:27]: I would have liked improve upon this, but with the current structure I find this necessary.

**Review by paschalidi — APPROVED** [review]
Thanks for putting the effort on this. Looks good to me. Lets get this in!

**Comment by JoakimSM** [comment]
Will merge part2 of this directly into v36 (rebased part2 on top of this)

**Pull request closed without merging**